### PR TITLE
[TOPI] Add tensor multiplication.

### PR DESCRIPTION
**Pull request by Vooblin**
Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/dmlc/tvm/blob/master/CONTRIBUTORS.md#reviewers).

### CODE REVIEW

**Comment by tqchen** [comment]
Thanks for the contribution! Please request reviews from reviewers

**Comment by Vooblin** [comment]
@nishi-t @were Could you please help in reviewing.

File: topi/python/topi/transform.py [hunk 1]
```diff
@@ -269,3 +269,22 @@ def matmul(a, b, transp_a=False, transp_b=False):
...
+    """A generalization of matrix multiplication to tensor.
+
+    `Res[i_1, ... , j_1, ...] = Sum_{k_1, ...} A[i_1 ..., k_1, ...]*B[k_1, ..., j_1, ...]`
+    The number of `k` variables is ndims_to_reduce.
+
+    Parameters
+    ----------
+    a : The tensor A
+    b : The tensor B
+    ndims_to_reduce : The number of dimensions to reduce over
```
Comment (nishi-t) [inline topi/python/topi/transform.py]: Could you change to use `axis` as with [numpy's tensordot](https://docs.scipy.org/doc/numpy-1.15.1/reference/generated/numpy.tensordot.html)?
I think It is better to make it compatible with numpy if possible.

**Comment by Vooblin** [comment]
@nishi-t thanks for review, I've updated patch. Now it fully compatible with numpy's tensordot. Could you help in reviewing again?

**Comment by nishi-t** [comment]
@Vooblin Sorry for the late reply. I'll review this in a few days.

File: topi/include/topi/transform.h [hunk 2]
```diff
@@ -745,6 +745,129 @@ inline tvm::Tensor matmul(const tvm::Tensor& A,
...
+    [](const std::vector<int>& vec, int elem) {
+      for (auto it = vec.begin(); it != vec.end(); ++it)
+        if (*it == elem)
+          return it;
+      return vec.end();
+    };
+
+  Array<Expr> output_shape;
+  for (unsigned i = 0; i < A->shape.size(); ++i)
+    if (vec_find(A_axes_val, i) == A_axes_val.end())
```
Comment (nishi-t) [inline topi/include/topi/transform.h]: Could you use `std::find` instead of `vec_find`?

**Comment by tqchen** [comment]
Thanks, @Vooblin @nishi-t , this is merged!

**Pull request merged**